fix: Update transaction validation logic in ethclient

Leekyungun opened the pull request:
# Description

This PR resolves an issue where blocks containing system transactions could not be retrieved properly using `getBlock`.  
For such blocks, the `transactionsRoot` is set to `EmptyTxsHash`, but the `Transactions` list contains one transaction, which leads to a mismatch and prevents block information from being retrieved.  
The validation logic has been updated to handle these cases correctly.
Example block with system transaction: https://amoy.polygonscan.com/block/16653120

# Changes

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] Changes only for a subset of nodes

# Breaking changes

No breaking changes have been introduced in this PR.

# Nodes audience

This PR affects all nodes and does not include changes limited to a subset of nodes.

# Checklist

- [ ] I have added at least 2 reviewer or the whole pos-v1 team
- [ ] I have added sufficient documentation in code
- [ ] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply
- [ ] Created a task in Jira and informed the team for implementation in Erigon client (if applicable)
- [ ] Includes RPC methods changes, and the Notion documentation has been updated

# Cross repository changes

- [ ] This PR requires changes to heimdall
    - In case link the PR here:
- [ ] This PR requires changes to matic-cli
    - In case link the PR here:

## Testing

- [ ] I have added unit tests
- [ ] I have added tests to CI
- [ ] I have tested this code manually on local environment
- [ ] I have tested this code manually on remote devnet using express-cli
- [ ] I have tested this code manually on mumbai/amoy
- [ ] I have created new e2e tests into express-cli

### Manual tests

1. Retrieved a block with system transactions (`transactionsRoot` as `EmptyTxsHash` but including `Transactions`).
2. Verified that block information is now retrieved correctly without validation errors.
3. Confirmed that other blocks without system transactions remain unaffected.

# Additional comments

See the example block for context:  
https://amoy.polygonscan.com/block/16653120

manav2401 commented:
Hi, thanks for raising this PR. The issues seems correct that it will happen when there's only 1 transaction which is a state-sync (system) transaction and nothing else. 

But, IMO, we shouldn't get rid of the error. Let me see if there's some way to find out if it's a state-sync tx or not.

manav2401 commented:
Hi @Leekyungun, can you update the PR to handle this specific condition explicitly instead of removing this check?

Basically, system transactions are directly applied on state and hence the `TxHash` field in header is unaffected (`EmptyTxsHash` in this case). Can you add a check to match the `to` and `from` fields of the only transaction. If it's `0x00..00`, then it's indeed a system transaction and we can skip this check in that case. If it's not, this check should be applied. 

Also, could you raise the PR against `develop` branch instead? Lmk if you're upto do this. Thanks!

Leekyungun commented:
```
Can you add a check to match the to and from fields of the only transaction. If it's 0x00..00, then it's indeed a system transaction and we can skip this check in that case.
```
Good idea!
I'll comment when the work is done.Thanks

Leekyungun commented:
I've added exception handling and updated the target branch to develop.

Initially, I considered declaring the address "0x0000000000000000000000000000000000000000" as a constant and replacing its occurrences, but I was concerned that future updates in Ethereum might cause conflicts.
For now, I’ve made minimal modifications.

Please provide feedback.
@manav2401

kamuikatsurgi commented on ethclient/ethclient.go:
```diff
@@ -213,6 +217,29 @@ func (ec *Client) getBlock(ctx context.Context, method string, args ...interface
+	zeroAddress := "0x0000000000000000000000000000000000000000"
+
+	// Check both 'from' and 'to' addresses
+	if transaction.From.String() == zeroAddress {
+		if transaction.tx.To() != nil && transaction.tx.To().String() == zeroAddress {
+			return true
+		}
+	}
```
you can combine the if checks condition here into a single one like this:
```go
if transaction.From.String() == zeroAddress && transaction.tx.To() != nil && transaction.tx.To().String() == zeroAddress {
```

kamuikatsurgi commented on ethclient/ethclient.go:
```diff
@@ -213,6 +217,29 @@ func (ec *Client) getBlock(ctx context.Context, method string, args ...interface
+// - The "from" and "to" addresses in the transaction are both zero addresses.
+func IsSystemTransaction(block rpcBlock) bool {
+	// Empty condition: Block must contain exactly one transaction to qualify
+	if len(block.Transactions) != 1 {
```
@manav2401 Can there be more than one system transaction? If yes, we should loop over all the transactions and ensure they are all system transactions.

Leekyungun replied:
I think checking all transactions would provide better scalability.
I’ll make the necessary changes and provide comments afterward.

kamuikatsurgi reviewed (commented):
Hey @Leekyungun, thanks for the PR, just small suggestions and would wait for final review from @manav2401 😄

Leekyungun commented:
Thanks for the review, it seems like the code has improved because of it. 
Please check it one more time.
@kamuikatsurgi

kamuikatsurgi commented:
> Thanks for the review, it seems like the code has improved because of it. Please check it one more time. @kamuikatsurgi

Hey, the PR loogs good, would wait for final review from @manav2401 😄

kamuikatsurgi commented:
> Thanks for the review, it seems like the code has improved because of it. Please check it one more time. @kamuikatsurgi

Hey, the PR looks good, wait for the final review from @manav2401 😄

manav2401 commented:
@Leekyungun thanks for updating the PR. I think it can be done in a simple way without introducing new functions. 

```go
if head.TxHash == types.EmptyTxsHash && len(body.Transactions) > 0 {
		tx := body.Transactions[0]
		zeroAddress := common.HexToAddress("0x0000000000000000000000000000000000000000")
		if *tx.From != zeroAddress || *tx.tx.To() != zeroAddress {
			return nil, errors.New("server returned non-empty transaction list but block header indicates no transactions")
		}
	}
```

Moreover, you can define `zeroAddress` in `bor_ethclient.go` as a public variable as it's bor specific.

Also, to answer your question, state sync transactions are always added at the end and there's exactly one of them. You can assume this for your implementation. Thanks.

Leekyungun commented:
```
        if head.TxHash == types.EmptyTxsHash && len(body.Transactions) > 0 {
		tx := body.Transactions[0]
		zeroAddress := common.HexToAddress("0x0000000000000000000000000000000000000000")
                 if (tx.From != nil && *tx.From != zeroAddress) || (tx.tx.To() != nil && *tx.tx.To() != zeroAddress) {
			return nil, errors.New("server returned non-empty transaction list but block header indicates no transactions")
		}
	}
```

Just to be safe, if tx.From and tx.tx.To() are both nil, it would cause a panic. How about adding a nil check?
@manav2401

manav2401 commented:
@Leekyungun it works. thanks!

Leekyungun commented:
I updated the implementation. 
Finally, I defined a Sync Transaction as a case where either from or to is zeroAddress. 
Can you verify if it should check for both being zeroAddress instead? 

Thank you!
@manav2401

manav2401 commented on ethclient/ethclient.go:
```diff
@@ -163,7 +163,13 @@ func (ec *Client) getBlock(ctx context.Context, method string, args ...interface
 	if head.TxHash == types.EmptyTxsHash && len(body.Transactions) > 0 {
-		return nil, errors.New("server returned non-empty transaction list but block header indicates no transactions")
+		// If there is only one transaction in the block and the header's txHash is `EmptyTxsHash`,
+		// it indicates a sync transaction. No error handling is required in this case.
```
```suggestion
		// it indicates a state-sync transaction. No error handling is required in this case.
```

manav2401 commented on ethclient/ethclient.go:
```diff
@@ -163,7 +163,13 @@ func (ec *Client) getBlock(ctx context.Context, method string, args ...interface
+		// If there is only one transaction in the block and the header's txHash is `EmptyTxsHash`,
+		// it indicates a sync transaction. No error handling is required in this case.
+		tx := body.Transactions[0]
+		zeroAddress := common.HexToAddress("0x0000000000000000000000000000000000000000")
```
Move this to `bor_ethclient.go` as it's bor specific.

manav2401 commented:
@Leekyungun current logic is fine. Few minor nits and then it's good to go.

Leekyungun commented:
Completed!
Please review it.
@manav2401

manav2401 commented:
hey @Leekyungun I messed something up locally. Can you create a fresh PR the changes required. 

Also, while you're at it, you can do the change I did [here](https://github.com/maticnetwork/bor/commit/a48254c82ebedca8b6fb1f7770f07a3c384c3e79). Apologies for this. Thanks!

Leekyungun commented:
It's OK! 
This is new PR: https://github.com/maticnetwork/bor/pull/1435
Rather, the commits are cleaner and better.

Please review it 
@manav2401 @kamuikatsurgi
